### PR TITLE
Gutenboarding - Add editor overrides + "Launch" Button

### LINKED ISSUE #38904: Gutenboarding: inject "Launch" button to Gutenberg header
For un-launched sites, we'll want to have "launch" button all the time visible that starts the "pre-launch flow" (plan selection, checkout, etc).

Design:

image

The live header looks like this when editing pages/posts (also for un-launched sites):

image

(you can ignore the pink color and imagine it's blue)

It will likely look different when editing _sites_, but even then there will likely be "Save/Update"  buttons to save edited site layout/global styles whatnot.

Curious to hear from designers how launch button should integrate with other buttons in the header?

Meanwhile, we can inject this button already to get into building a functional prelaunch flow for Gutenboarding. It should stay hidden behind a flag.

cc @mtias @olesyabrk @joanrho @dwolfe for thoughts.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
#### Changes proposed in this Pull Request

Adds an optional override to update the wpcom-block-editor for gutenboarding purposes.

Currently, this override will happen if both of two conditions are `true`:

1.  `config.isEnabled( 'gutenboarding' )` - Which is currently true only in development and wpcalypso (calypso.live) environments.
2. `is-gutenboarding` query param is present.

When enabled, the settings bar in the top right of the editor is changed from:
Screen Shot 2020-01-24 at 4 25 33 PM

To:
Screen Shot 2020-01-24 at 4 32 52 PM

* The "Save" button is currently a re-styling of the original "Update" or "Publish" button.
* The "Launch" button is wrapped in an <a> tag leading to the frankenflow.

#### Testing instructions

* Run this PR on calypso.
* Build and Sync wpcom-block-editor to your sandbox.  You can patch `D38314` or alternatively run `./bin/calypso/build-wpcom-block-editor.sh 583382` on your sandbox, where '583382' represents the circle-ci build number for the wpcom-block-editor found at the bottom of this PR. 
* Be sure to sandbox `widgets.wp.com` and `public-api.wordpress.com`.
* Navigate to the editor in a sandboxed site and verify there is NO "Launch" button.  
* Add the query param `?is-gutenboarding` to the end of the url in your browser window and load the editor again.
* Verify there are "Save" and "Launch" buttons, and that the styles match with the image shown above. 
 You may need to clear your browser's cache to see changes.
* Verify the "Launch" button leads to the first step of the Frankenflow.

Closes #38904

### CODE REVIEW

File: apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js [hunk 1]
```diff
@@ -679,6 +679,24 @@ function getCloseButtonUrl( calypsoPort ) {
 	};
 }
 
+// TODO - JSDOC this when you know what you are actually needing to do...
+function getLaunchButton( calypsoPort ) {
+	const { port1, port2 } = new MessageChannel();
```
Comment (jeraldjuice) [inline apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: `new window.MessageChannel()` here will fix the linter error :)

**Comment by sirreal** [comment]
> Oof this PR got hit with the `any` crackdown, re: TS linter failures. iirc, probably not worth the time fixing in tandem with this issue. Possibly the best bet is to just to add an ignore for the rule in the calyposify iframe if it's blocking your build.

Chipping away a bit at this, but I agree, we'll need to ignore lint issues.

**Comment by Addison-Stavlo** [comment]
Some notes on what is working and what will be required (at minimum) before being ready to merge.

## First - The bare minimum needed before this can be merged.
The logic for when these changes to the editor should be in effect needs to be updated.  We can add this logic to our `GetGutenboardingStatus` case of the dispatch table in `client/gutenberg/editor/calypsoify-iframe.tsx`.  Currently, this is only set to check if gutenboarding is enabled in config meaning that any instance of the editor opened in a development environment will have these changes.  This is not desired, and we will need to add some more logic about the site in question and its status in the onboarding process here before we can think about merging this PR.

There are definitely other aspects of this that will need to be updated, but from what I can tell this is currently the only large blocker for the first iteration.

## What is working?
**TLDR**:  Communication between both sides of the iframe works, and the Launch button injection/override takes place using this information.

**More on the current process**
When it begins to load, the bridge-server sends a message that is received by the calypsoify-iframe which then checks the gutenboarding logic and responds back to the bridge-server with a package of data.  Currently this data contains a boolean value for `isGutenboarding` and a rough cut of the `frankenflowUrl` to send the user to when the Launch button is clicked.  The bridge-server unpacks this data onto the `calypsoifyGutenberg` object for later use by the `launch-button-override` file, and triggers a hook that is only necessary as a fallback in the case that this data is unpacked after the editor has already been rendered.

The `launch-button-override` file runs on `domReady`.  If `isGutenboarding` is truthy on the `calypsoifyGutenberg` object, then it runs a function to update the button bar.  A hook is defined to run this function as well (as noted before this is a fallback in the case that `isGutenboarding` has not been set by the time `domReady` runs).  From my current testing, this has always been set and the hook has never needed to run.  However, this hook was (for some reason) seemingly necessary when similar work was done in the dotcom FSE editor which is why I have left it here as a fallback for now.

The `updateButtonBar` function makes some simple updates to the button bar:
* Hides the 'switch to draft' button.
* Hides the 'Preview' button.
* Changes the "Update"/"Publish" primary button to a tertiary button that says "save".
* Adds a primary "Launch" button wrapped in an anchor tag leading to the frankenflow.

**Before `updateButtonBar`**
Screen Shot 2020-01-24 at 4 25 33 PM

**After `updateButtonBar`**
Screen Shot 2020-01-24 at 4 32 52 PM


There will still be some bugs to work out, and we may want to redesign how some things are handled (how we launch into frankenflow, how the buttons are replaced or removed, etc.), but I think this gives us a good starting point to be able to send information from calypso that allows us to conditionally change certain pieces of the editor for gutenboarding.

File: apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js [hunk 2]
```diff
@@ -679,6 +680,31 @@ function getCloseButtonUrl( calypsoPort ) {
...
+			payload: {},
+		},
+		[ port2 ]
+	);
+	port1.onmessage = ( { data } ) => {
+		const { isGutenboarding, frankenflowUrl } = data;
+		calypsoifyGutenberg.isGutenboarding = isGutenboarding;
+		calypsoifyGutenberg.frankenflowUrl = frankenflowUrl;
+		// Hook necessary if message recieved after editor has loaded.
+		window.wp.hooks.doAction( 'updateLaunchButton', isGutenboarding );
```
Comment (noahtallen) [inline apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: Do you think it makes sense to switch to creating an @wordpress/data store? Then, it could update state through that instead of the option.

Theoretically, I'd love to switch the FSE button to the same mechanism if possible. It might be more effort than we need to put in at this point, but it might be useful if there is more data we want to think about for the state here

Comment (Addison-Stavlo) [reply apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: Thats about what I thought as well.  The @wordpress/data store would work well, but since we have the calypsoifyGutenberg object available and only need to tack on two static values it didn't seem worth setting up yet.  That could definitely be something worth doing in the end though, especially if we end up needing to handle more state.

Comment (jeraldjuice) [reply apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: Yeah, I agree with @Addison-Stavlo on this one - I ran into a similar thing with the background color PR (#39045). I think once we get further along with frankenflow, we'll have something in the store to reference, opposed to adding an arbitrary one now.

Comment (noahtallen) [reply apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js]: That makes a ton of sense! :)

**Comment by Addison-Stavlo** [comment]
I figured I should post a comment since my last one was during WIP.

But we should be good to go from here for testing, review, and if all looks good then merging.

**Comment by sirreal** [comment]
Rebased.

**Comment by sirreal** [comment]
I've prepared D38286-code from the HEAD commit for ease of testing.

**Comment by sirreal** [comment]
I've been trying to sandbox and test this with D38286-code. The plugin seems to be running, but I never see the Launch button.

**Comment by Addison-Stavlo** [comment]
@sirreal  -  It doesn't look like that diff includes the output from the `build-wpcom-block-editor.sh` process?  I ran that diff as well and coudn't find the launch button.  However, running `./bin/calypso/build-wpcom-block-editor.sh 583016` (583016 being the build number from ci/circleci: build-wpcom-block-editor below) in the sandbox seems to make the changes apparent.

**Comment by sirreal** [comment]
> […] running `./bin/calypso/build-wpcom-block-editor.sh 583016` […] seems to make the changes apparent.

That works much better, thanks! I also need to sandbox `widgets.wp.com` which wasn't obvious to me initially.

File: apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+		// Create a 'Launch' button.
+		const launchButton = document.createElement( 'button' );
+		launchButton.className =
+			'gutenboarding-editor-overrides__launch-button components-button is-primary';
+		launchButton.innerText = __( 'Launch' );
+
+		// Wrap 'Launch' button in anchor to frankenflow.
+		const launchButtonWrapper = document.createElement( 'a' );
+		launchButtonWrapper.href = calypsoifyGutenberg.frankenflowUrl;
```
Comment (sirreal) [inline apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js]: We end up navigating the iframe to the URL, which isn't really what we want. We'd like navigation to happen in Calypso.

We should be able to get that by setting the a `target` to `_top`.

```suggestion
		launchButtonWrapper.href = calypsoifyGutenberg.frankenflowUrl;
		launchButtonWrapper.target = '_top';
```

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target

File: client/gutenberg/editor/calypsoify-iframe.tsx [hunk 4]
```diff
@@ -279,6 +280,17 @@ class CalypsoifyIframe extends Component< Props & ConnectedProps & ProtectedForm
...
 		}
 
+		if ( EditorActions.GetGutenboardingStatus === action ) {
+			// TODO - In future iterations, replace window param with gutenboarding site info.
+			const urlParams = new URLSearchParams( window.location.search );
+			const isGutenboarding =
+				config.isEnabled( 'gutenboarding' ) && urlParams.has( 'is-gutenboarding' );
+			ports[ 0 ].postMessage( {
+				isGutenboarding,
+				frankenflowUrl: `${ window.location.origin }/start/frankenflow`,
```
Comment (sirreal) [inline client/gutenberg/editor/calypsoify-iframe.tsx]: We're supposed to provide the `siteSlug` for frankenflow. Site ID is usually interchangeable and this seems to work:

```suggestion
				frankenflowUrl: `${ window.location.origin }/start/frankenflow?siteSlug=${ this.props.siteId }`,
```

**Comment by Addison-Stavlo** [comment]
> That works much better, thanks! I also need to sandbox widgets.wp.com which wasn't obvious to me initially.

Gotcha.  I updated the testing instructions to try to make that more apparent.  I think `widgets.wp.com` stands out a bit better than just the word "widgets" in the wall of testing instructions.

File: apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js (hunk 3 above)
Comment (Addison-Stavlo) [reply apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js]: Great catch, great fix.  Going to test that now.

File: client/gutenberg/editor/calypsoify-iframe.tsx (hunk 4 above)
Comment (Addison-Stavlo) [reply client/gutenberg/editor/calypsoify-iframe.tsx]: sounds good!

**Comment by Addison-Stavlo** [comment]
> This is great, it's really close! I've left a couple small suggestions.

Great suggestions.  They have been applied so now it goes into frankenflow outside of the iframe and with the site slug.

**Comment by Addison-Stavlo** [comment]
Created diff `D38314` that will need to be reviewed and shipped along with this PR.

**Comment by sirreal** [comment]
I've rebased and merged #39057 here.

**Comment by sirreal** [comment]
> Created diff D38314-code that will need to be reviewed and shipped along with this PR.

I've updated D38314-code from job https://circleci.com/gh/Automattic/wp-calypso/584360 / 216efcdde31dfd86bbfc695c032be44f08b913e8 (after merge/rebase)

File: apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			return;
+		}
+		clearInterval( awaitSettingsBar );
+
+		// 'Update'/'Publish' primary button to become 'Save' tertiary button.
+		const saveButton = settingsBar.querySelector( '.editor-post-publish-button' );
+		saveButton && ( saveButton.innerText = __( 'Save' ) );
+
+		// Create a 'Launch' button.
+		const launchButton = document.createElement( 'button' );
```
Comment (simison) [inline apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js]: We end up with `a` _and_ `button`:
```html
<a href="...">
  <button class="gutenboarding-editor-overrides__launch-button components-button is-primary">
    Launch
  </button>
</a>
```

...which isn't a biggie but it has little funky visual hover state (note underline):

image

A simple `<a href="..." class="...">Launch</a>` would look normal:

image

Comment (sirreal) [reply apps/wpcom-block-editor/src/calypso/features/gutenboarding-editor-overrides.js]: #39167 👈
